source/common/router/router.cc

```diff
@@ -515,44 +555,119 @@ void Filter::onRequestComplete() {
 }
 
 void Filter::onDestroy() {
-  if (upstream_request_ && !attempting_internal_redirect_with_complete_stream_) {
-    upstream_request_->resetStream();
+  if (!attempting_internal_redirect_with_complete_stream_) {
+    for (unsigned long i = 0; i < upstream_requests_.size(); i++) {
+      if (upstream_requests_[i]) {
+        upstream_requests_[i]->resetStream();
+      }
+    }
   }
   cleanup();
 }
 
 void Filter::onResponseTimeout() {
   ENVOY_STREAM_LOG(debug, "upstream timeout", *callbacks_);
-  cluster_->stats().upstream_rq_timeout_.inc();
 
-  // It's possible to timeout during a retry backoff delay when we have no upstream request.
-  if (upstream_request_) {
-    if (upstream_request_->upstream_host_) {
-      upstream_request_->upstream_host_->stats().rq_timeout_.inc();
+  // Reset any upstream requests that are still in flight.
+  for (unsigned long i = 0; i < upstream_requests_.size(); i++) {
+    // Don't record a timeout for upstream requests we've already seen headers
+    // for.
+    UpstreamRequest* upstream_request = upstream_requests_[i].get();
+    if (upstream_request && !upstream_request->upstream_headers_) {
+      cluster_->stats().upstream_rq_timeout_.inc();
+      if (upstream_request->upstream_host_) {
+        upstream_request->upstream_host_->stats().rq_timeout_.inc();
+      }
+
+      // If this upstream request already hit a "soft" timeout, then it
+      // already recorded a timeout into outlier detection. Don't do it again.
+      if (!upstream_request->outlier_detection_timeout_recorded_) {
+        updateOutlierDetection(timeout_response_code_, upstream_request);
+      }
+      upstream_request->resetStream();
+
+      chargeUpstreamAbort(timeout_response_code_, false, upstream_request);
     }
-    upstream_request_->resetStream();
   }
 
-  updateOutlierDetection(timeout_response_code_);
   onUpstreamTimeoutAbort(StreamInfo::ResponseFlag::UpstreamRequestTimeout);
 }
 
-void Filter::onPerTryTimeout() {
-  updateOutlierDetection(timeout_response_code_);
+// Called when the per try timeout is hit but we didn't reset the request
+// (hedge_on_per_try_timeout enabled).
+void Filter::onSoftPerTryTimeout(UpstreamRequest* upstream_request) {
+  // Even though we didn't cancel the request yet we still want to track it
+  // in outlier detection.
+  // TODO(mpuncel) is it weird to have a pretend response code here? we might
+  // get a 200 back from this request later.
+  updateOutlierDetection(timeout_response_code_, upstream_request);
```

Comment:
could use feedback on this. I think it's good for the outlier detection to know that this host timed out, but it's weird that we might save 2 HTTP codes in for a single request (504 after soft timeout, and then possible 5xx later on).

Reply:
Given that outlier detection is about tracking responses from individual hosts I think tracking both is fine